Added NAT-PMP, PCP, UPnP probing functionality

kennysong opened the pull request:
Probes the router for these port control protocols on the "Logs & Network Analysis" page. 

Solves #1604.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/1630)
<!-- Reviewable:end -->

bemasc commented on src/generic_core/diagnose-nat.ts:
```diff
@@ -854,3 +856,310 @@ export function doNatProvoking() :Promise<string> {
 }
+
+// Test if NAT-PMP is supported by the router
+export function probePMPSupport(routerIP :string) :Promise<string> {
```
I think it would make more sense for this function to return `Promise<boolean>` (`true` means supported) or `Promise<void>` (resolve means supported).

bemasc commented:
Review status: 0 of 4 files reviewed, 10 unresolved discussions, all commit checks successful.

---

<sup>**[src/chrome/app/manifest.json, line 18 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrSyxrz8Gjqh1VidZu-)** ([raw file](https://github.com/uproxy/uproxy/blob/a05e4b7bc28d994739c37adc1867ada394a4b341/src/chrome/app/manifest.json#L18)):</sup>
Adding new permissions, especially broad ones like this, is not very appealing.  I think you should be able to avoid the need to add this permission by using `freedom['core.xhr']`instead of `XMLHttpRequest`.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 862 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrSzQ7D6VvNjJDoCTIg)** ([raw file](https://github.com/uproxy/uproxy/blob/a05e4b7bc28d994739c37adc1867ada394a4b341/src/generic_core/diagnose-nat.ts#L862)):</sup>
There's no need to wrap this in a function.  You can just compute the promise, and then drop that into Promise.race() at the end.  Same for the other functions.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 867 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrT-BOKqZDPy14kWNHL)** ([raw file](https://github.com/uproxy/uproxy/blob/a05e4b7bc28d994739c37adc1867ada394a4b341/src/generic_core/diagnose-nat.ts#L867)):</sup>
Our style guide generally requires that ordinary variables start with lower case, so this would be `pmpResponse`.  (We make an exception for `F` and `R`.)

---

<sup>**[src/generic_core/diagnose-nat.ts, line 875 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrSzf--rM-IXO4-rCoq)** ([raw file](https://github.com/uproxy/uproxy/blob/a05e4b7bc28d994739c37adc1867ada394a4b341/src/generic_core/diagnose-nat.ts#L875)):</sup>
This should be a reject, not a resolve.  Same for the other functions.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 935 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrT1TCa-Hgc2JiTjb3o)** ([raw file](https://github.com/uproxy/uproxy/blob/a05e4b7bc28d994739c37adc1867ada394a4b341/src/generic_core/diagnose-nat.ts#L935)):</sup>
Float?

---

<sup>**[src/generic_core/diagnose-nat.ts, line 939 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrT1lG5d8pt8wAPOQZ1)** ([raw file](https://github.com/uproxy/uproxy/blob/a05e4b7bc28d994739c37adc1867ada394a4b341/src/generic_core/diagnose-nat.ts#L939)):</sup>
I think this is a good place to use ipaddr.js

---

<sup>**[src/generic_core/diagnose-nat.ts, line 1027 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrT2eQ-E1t0LAAfDHtz)** ([raw file](https://github.com/uproxy/uproxy/blob/a05e4b7bc28d994739c37adc1867ada394a4b341/src/generic_core/diagnose-nat.ts#L1027)):</sup>
"get" is a bit ambiguous.  How about "fetch"?

---

<sup>**[src/generic_core/diagnose-nat.ts, line 1135 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrT129yMK5s0TIqOQNa)** ([raw file](https://github.com/uproxy/uproxy/blob/a05e4b7bc28d994739c37adc1867ada394a4b341/src/generic_core/diagnose-nat.ts#L1135)):</sup>
I think this is overkill.  I would suggest just splitting on whitespace and pulling out the token at the appropriate index, then using ipaddr.js to determine if it's IPv4.

---

<sup>**[third_party/typings/generic/url.d.ts, line 22 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrT3IJZCpk3kyhA_zPT)** ([raw file](https://github.com/uproxy/uproxy/blob/a05e4b7bc28d994739c37adc1867ada394a4b341/third_party/typings/generic/url.d.ts#L22)):</sup>
Missing newline.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1630)
<!-- Sent from Reviewable.io -->

lucyhe commented:
lgtm, awesome work! I'm also just a little worried by the broad http permission in the manifest -- if we can get around that, that'd be ideal.

---

Review status: 3 of 4 files reviewed, 10 unresolved discussions, all commit checks successful.
Reviewed files:
- src/chrome/app/manifest.json @ r1
- src/generic_core/uproxy_core.ts @ r1
- third_party/typings/generic/url.d.ts @ r1

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1630)
<!-- Sent from Reviewable.io -->

willscott commented:
agree that the http://*/ permission is a bit scary.  Could make that go away by making the XHR through sockets. there are libraries to help with that exactly to avoid this sort of permission problem: https://github.com/ahmadnassri/chrome.sockets.tcp.xhr

bemasc commented:
Actually ... maybe it doesn't matter?  We should check whether it alters the prompt to the user.  Arguably, it should not, because the app already requests full TCP permission, so "http://*/" is not actually any additional privilege.

kennysong commented:
The permissions prompt with `http://*/` is the same:

```
 - Communicate with cooperating websites
 - Exchange data with any device on the local network or internet
```

---

Review status: 3 of 4 files reviewed, 10 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1630)
<!-- Sent from Reviewable.io -->

bemasc commented:
What about Firefox?

---

Review status: 3 of 4 files reviewed, 10 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1630)
<!-- Sent from Reviewable.io -->

kennysong commented:
I only changed `manifest.json`, which only applies to Chrome. The new probing functionality works in Firefox without any extra permissions.

---

Review status: 3 of 4 files reviewed, 10 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1630)
<!-- Sent from Reviewable.io -->

kennysong commented:
Review status: 1 of 4 files reviewed, 10 unresolved discussions, some commit checks pending.

---

<sup>**[src/chrome/app/manifest.json, line 18 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrSyxrz8Gjqh1VidZu-)** ([raw file](https://github.com/uproxy/uproxy/blob/a05e4b7bc28d994739c37adc1867ada394a4b341/src/chrome/app/manifest.json#L18)):</sup>
Didn't change since this doesn't alter the user permission prompts in Chrome or Firefox. (See discussion above)

---

<sup>**[src/generic_core/diagnose-nat.ts, line 861 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrSwyjLdCQ1KCi0TUsa-r1-861)** ([raw file](https://github.com/uproxy/uproxy/blob/a05e4b7bc28d994739c37adc1867ada394a4b341/src/generic_core/diagnose-nat.ts#L861)):</sup>
Done. `probePMPSupport()` and `probePCPSupport()` now resolve to booleans, while `probeUPnPSupport` resolves only to a `true` boolean (if everything goes smoothly) or rejects with an Error containing a detailed message. Booleans and Errors are converted into status strings in the wrapper functions in `uproxy_core.ts`.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 862 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrSzQ7D6VvNjJDoCTIg)** ([raw file](https://github.com/uproxy/uproxy/blob/a05e4b7bc28d994739c37adc1867ada394a4b341/src/generic_core/diagnose-nat.ts#L862)):</sup>
Done.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 867 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrT-BOKqZDPy14kWNHL)** ([raw file](https://github.com/uproxy/uproxy/blob/a05e4b7bc28d994739c37adc1867ada394a4b341/src/generic_core/diagnose-nat.ts#L867)):</sup>
Should be fixed here and in other variables.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 875 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrSzf--rM-IXO4-rCoq)** ([raw file](https://github.com/uproxy/uproxy/blob/a05e4b7bc28d994739c37adc1867ada394a4b341/src/generic_core/diagnose-nat.ts#L875)):</sup>
Done.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 935 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrT1TCa-Hgc2JiTjb3o)** ([raw file](https://github.com/uproxy/uproxy/blob/a05e4b7bc28d994739c37adc1867ada394a4b341/src/generic_core/diagnose-nat.ts#L935)):</sup>
It was slightly more compact than two `setInt32()`'s (there's no `setInt64()`). I changed it to the former for clarity, though.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 939 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrT1lG5d8pt8wAPOQZ1)** ([raw file](https://github.com/uproxy/uproxy/blob/a05e4b7bc28d994739c37adc1867ada394a4b341/src/generic_core/diagnose-nat.ts#L939)):</sup>
Done.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 1027 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrT2eQ-E1t0LAAfDHtz)** ([raw file](https://github.com/uproxy/uproxy/blob/a05e4b7bc28d994739c37adc1867ada394a4b341/src/generic_core/diagnose-nat.ts#L1027)):</sup>
Done.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 1135 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrT129yMK5s0TIqOQNa)** ([raw file](https://github.com/uproxy/uproxy/blob/a05e4b7bc28d994739c37adc1867ada394a4b341/src/generic_core/diagnose-nat.ts#L1135)):</sup>
Done.

---

<sup>**[third_party/typings/generic/url.d.ts, line 22 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrT3IJZCpk3kyhA_zPT)** ([raw file](https://github.com/uproxy/uproxy/blob/a05e4b7bc28d994739c37adc1867ada394a4b341/third_party/typings/generic/url.d.ts#L22)):</sup>
Done.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1630)
<!-- Sent from Reviewable.io -->

bemasc commented:
Review status: 1 of 4 files reviewed, 11 unresolved discussions, all commit checks successful.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 909 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXsqiR3agFR2cyjAfc)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/diagnose-nat.ts#L909)):</sup>
It looks to me like you are leaking this socket on every test.  Unfortunately, Freedom objects aren't garbage-collected, so you need to explicitly close them.  In fact, to be really clean, you should probably do
```typescript
socket.close().then(() => {
  freedom['core.udpsocket'].close(socket);
});
```
The first line closes the OS-level socket, and the second line discards the Freedom object representing that socket.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 917 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXsiR8C7qNUNjePj0N)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/diagnose-nat.ts#L917)):</sup>
Why not bind to `privateIP`?

---

<sup>**[src/generic_core/diagnose-nat.ts, line 1033 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXv7eAR734Y6iq9_j3)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/diagnose-nat.ts#L1033)):</sup>
We try to always use `===` for consistency.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 1078 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXvIQOYIgnvqQ5ivzR)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/diagnose-nat.ts#L1078)):</sup>
Two spaces before an inline comment

---

<sup>**[src/generic_core/diagnose-nat.ts, line 1104 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXsIDCEsiOrGMTn5WA)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/diagnose-nat.ts#L1104)):</sup>
Our style guide excludes inline `if { }`

---

<sup>**[src/generic_core/diagnose-nat.ts, line 1120 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXsZg8HtuQz4uTOjiA)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/diagnose-nat.ts#L1120)):</sup>
80 column limit?

---

<sup>**[src/generic_core/diagnose-nat.ts, line 1123 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXsWMaEI9Awqwl3LKl)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/diagnose-nat.ts#L1123)):</sup>
No space between `?` and `:`.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 1127 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrT129yMK5s0TIqOQNa)** ([raw file](https://github.com/uproxy/uproxy/blob/a05e4b7bc28d994739c37adc1867ada394a4b341/src/generic_core/diagnose-nat.ts#L1127)):</sup>
Still overkill :).  Just hardcode the value of `hostIndex` and check if what's there is `"host"`.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 1138 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXvcQzQ3qye_sRrnKQ)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/diagnose-nat.ts#L1138)):</sup>
4 space indent

---

<sup>**[src/generic_core/diagnose-nat.ts, line 1156 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXuxe-TgahJInYvawf)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/diagnose-nat.ts#L1156)):</sup>
No space before `:` in arguments

---

<sup>**[src/generic_core/uproxy_core.ts, line 374 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXwNKHD_stCSEwuOV_)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/uproxy_core.ts#L374)):</sup>
Rather than copy-paste this code between `probePMP` and `probePCP`, I think it would be better to have a single function that takes `probePCPSupport` or `probePMPSupport` as an argument.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1630)
<!-- Sent from Reviewable.io -->

kennysong commented:
Review status: 1 of 4 files reviewed, 11 unresolved discussions, some commit checks pending.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 909 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXsqiR3agFR2cyjAfc)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/diagnose-nat.ts#L909)):</sup>
Done.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 917 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXsiR8C7qNUNjePj0N)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/diagnose-nat.ts#L917)):</sup>
Done. Changed all the probing functions to bind to `privateIP` for consistency.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 1033 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXv7eAR734Y6iq9_j3)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/diagnose-nat.ts#L1033)):</sup>
Done. (Changed everywhere)

---

<sup>**[src/generic_core/diagnose-nat.ts, line 1078 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXvIQOYIgnvqQ5ivzR)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/diagnose-nat.ts#L1078)):</sup>
Done.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 1104 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXsIDCEsiOrGMTn5WA)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/diagnose-nat.ts#L1104)):</sup>
Didn't change. (Style guide doesn't actually exclude it)

---

<sup>**[src/generic_core/diagnose-nat.ts, line 1120 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXsZg8HtuQz4uTOjiA)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/diagnose-nat.ts#L1120)):</sup>
Done. Fixed a few other places as well.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 1123 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXsWMaEI9Awqwl3LKl)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/diagnose-nat.ts#L1123)):</sup>
Done.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 1127 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrT129yMK5s0TIqOQNa)** ([raw file](https://github.com/uproxy/uproxy/blob/a05e4b7bc28d994739c37adc1867ada394a4b341/src/generic_core/diagnose-nat.ts#L1127)):</sup>
Done.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 1138 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXvcQzQ3qye_sRrnKQ)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/diagnose-nat.ts#L1138)):</sup>
Done.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 1156 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXuxe-TgahJInYvawf)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/diagnose-nat.ts#L1156)):</sup>
Done.

---

<sup>**[src/generic_core/uproxy_core.ts, line 374 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXwNKHD_stCSEwuOV_)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/uproxy_core.ts#L374)):</sup>
Done.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1630)
<!-- Sent from Reviewable.io -->

bemasc commented:
Review status: 1 of 4 files reviewed, 5 unresolved discussions, some commit checks failed.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 909 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXsqiR3agFR2cyjAfc)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/diagnose-nat.ts#L909)):</sup>
What if there's no reply, so onData never fires?

---

<sup>**[src/generic_core/diagnose-nat.ts, line 1142 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXvcQzQ3qye_sRrnKQ)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/diagnose-nat.ts#L1142)):</sup>
That's a 2-space indent.

---

<sup>**[src/generic_core/uproxy_core.ts, line 396 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrYsLsffb8mjsUUneET)** ([raw file](https://github.com/uproxy/uproxy/blob/5270b9961e090f10f16829d446863eaed69bc112/src/generic_core/uproxy_core.ts#L396)):</sup>
Nit: Generally, we don't break lines in a case like this.  (`then` would go on the preceding line.)

---

<sup>**[src/generic_core/uproxy_core.ts, line 406 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrYsfci2lmO0VLZi5y6)** ([raw file](https://github.com/uproxy/uproxy/blob/5270b9961e090f10f16829d446863eaed69bc112/src/generic_core/uproxy_core.ts#L406)):</sup>
Printing the same error message three times seems unhelpful.

I think this would be a lot clearer if the success/failure handling were independent (before the Promise.all).

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1630)
<!-- Sent from Reviewable.io -->

kennysong commented:
Review status: 1 of 4 files reviewed, 5 unresolved discussions, all commit checks successful.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 909 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXsqiR3agFR2cyjAfc)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/diagnose-nat.ts#L909)):</sup>
Ah okay, I added a callback to destroy the socket when there's no reply/on timeout, which is wrapped in the new function `closeSocket()`.

---

<sup>**[src/generic_core/diagnose-nat.ts, line 1142 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrXvcQzQ3qye_sRrnKQ)** ([raw file](https://github.com/uproxy/uproxy/blob/06f08cc857bc7d5fc1ca7d9257b38a5a21180f55/src/generic_core/diagnose-nat.ts#L1142)):</sup>
Done.

---

<sup>**[src/generic_core/uproxy_core.ts, line 396 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrYsLsffb8mjsUUneET)** ([raw file](https://github.com/uproxy/uproxy/blob/5270b9961e090f10f16829d446863eaed69bc112/src/generic_core/uproxy_core.ts#L396)):</sup>
Done.

---

<sup>**[src/generic_core/uproxy_core.ts, line 406 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-JrYsfci2lmO0VLZi5y6)** ([raw file](https://github.com/uproxy/uproxy/blob/5270b9961e090f10f16829d446863eaed69bc112/src/generic_core/uproxy_core.ts#L406)):</sup>
The error handling is already managed independently in the wrapper functions `probePMCP()` and `probeUPnP()`; those should never reject and always return a status string. The `.catch()` here only applies to the error when `getInternalIP()` rejects.

However, I made the logic more sequential and compacted the three identical error messages into one.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1630)
<!-- Sent from Reviewable.io -->

bemasc commented:
Review status: 1 of 4 files reviewed, 3 unresolved discussions, all commit checks successful.

---

<sup>**[src/generic_core/uproxy_core.ts, line 368 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-Jrd50fzfkp_eoFe05aV)** ([raw file](https://github.com/uproxy/uproxy/blob/a3714fd7715ff09708a65e735c86e741957d0ba1/src/generic_core/uproxy_core.ts#L368)):</sup>
Nit: Don't put a line break here.

---

<sup>**[src/generic_core/uproxy_core.ts, line 383 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-Jrd54ygOfFhdAqK_7BN)** ([raw file](https://github.com/uproxy/uproxy/blob/a3714fd7715ff09708a65e735c86e741957d0ba1/src/generic_core/uproxy_core.ts#L383)):</sup>
No line break here.

---

<sup>**[src/generic_core/uproxy_core.ts, line 408 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-Jrd5Q7k_f1t6oZeA58p)** ([raw file](https://github.com/uproxy/uproxy/blob/a3714fd7715ff09708a65e735c86e741957d0ba1/src/generic_core/uproxy_core.ts#L408)):</sup>
I think it would be better to move `getNatType` before `getInternalIP`, so you don't have to call it twice.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1630)
<!-- Sent from Reviewable.io -->

kennysong commented:
Review status: 1 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

<sup>**[src/generic_core/uproxy_core.ts, line 368 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-Jrd50fzfkp_eoFe05aV)** ([raw file](https://github.com/uproxy/uproxy/blob/a3714fd7715ff09708a65e735c86e741957d0ba1/src/generic_core/uproxy_core.ts#L368)):</sup>
Done.

---

<sup>**[src/generic_core/uproxy_core.ts, line 383 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-Jrd54ygOfFhdAqK_7BN)** ([raw file](https://github.com/uproxy/uproxy/blob/a3714fd7715ff09708a65e735c86e741957d0ba1/src/generic_core/uproxy_core.ts#L383)):</sup>
Done.

---

<sup>**[src/generic_core/uproxy_core.ts, line 408 \[r4\]](https://reviewable.io:443/reviews/uproxy/uproxy/1630#-Jrd5Q7k_f1t6oZeA58p)** ([raw file](https://github.com/uproxy/uproxy/blob/a3714fd7715ff09708a65e735c86e741957d0ba1/src/generic_core/uproxy_core.ts#L408)):</sup>
Done.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1630)
<!-- Sent from Reviewable.io -->

bemasc commented:
:+1: 

---







Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1630)
<!-- Sent from Reviewable.io -->
